pkg/sysconfwatch/sysconfwatch.go

```diff
@@ -0,0 +1,132 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package sysconfwatch
+
+import (
+	"context"
+	"math"
+	"os"
+	"time"
+
+	"github.com/openshift/microshift/pkg/config"
+	"github.com/openshift/microshift/pkg/util"
+	"golang.org/x/sys/unix"
+	"k8s.io/klog/v2"
+)
+
+const sysConfigCheckInterval = time.Second * 5
+const sysConfigAllowedTimeDrift = time.Second * 10
+
+type SysConfWatchController struct {
+	NodeIP  string
+	timerFd int
+}
+
+func NewSysConfWatchController(cfg *config.MicroshiftConfig) *SysConfWatchController {
+	// Create a realtime clock timer with asynchronous read support
+	fd, err := unix.TimerfdCreate(unix.CLOCK_REALTIME, unix.TFD_CLOEXEC|unix.TFD_NONBLOCK)
+	if err != nil {
+		klog.Fatalf("failed to create a realtime clock timer %v", err)
+	}
+
+	// Set the time interval into distant future
+	var ptime = &unix.ItimerSpec{
+		Interval: unix.Timespec{Sec: math.MaxInt64, Nsec: 0},
+		Value:    unix.Timespec{Sec: 0, Nsec: 0},
+	}
+	// Start the timer with cancelation if the clock is reset elsewhere
+	err = unix.TimerfdSettime(fd, unix.TFD_TIMER_ABSTIME|unix.TFD_TIMER_CANCEL_ON_SET, ptime, nil)
+	if err != nil {
+		klog.Fatalf("failed to start a realtime clock timer %v", err)
+	}
+
+	return &SysConfWatchController{
+		NodeIP:  cfg.NodeIP,
+		timerFd: fd,
+	}
+}
+
+func (s *SysConfWatchController) Name() string { return "sysconfwatch-controller" }
+func (s *SysConfWatchController) Dependencies() []string {
+	return []string{}
+}
+
+func sysMonTimeDiff() int64 {
+	var stm unix.Timespec
+	var mtm unix.Timespec
+
+	// System-wide clock that measures real (i.e. wall-clock) time
+	// This clock is affected by discontinuous jumps in the system time (e.g. if the system administrator manually changes the clock)
+	// and by the incremental adjustments performed by adjtime and NTP
+	unix.ClockGettime(unix.CLOCK_REALTIME, &stm)
+	// Clock that cannot be set and represents monotonic time since some unspecified starting point
+	// It provides access to a raw hardware-based time that is not subject to NTP adjustments
+	// or the incremental adjustments performed by adjtime
+	unix.ClockGettime(unix.CLOCK_MONOTONIC_RAW, &mtm)
+
+	return stm.Sec - mtm.Sec
+}
+
+func (c *SysConfWatchController) Run(ctx context.Context, ready chan<- struct{}, stopped chan<- struct{}) error {
+	defer close(stopped)
+	ticker := time.NewTicker(sysConfigCheckInterval)
+	defer ticker.Stop()
+
+	klog.Infof("starting sysconfwatch-controller with IP address %q", c.NodeIP)
+
+	var buf []byte = make([]byte, 8)
+	// Take a snapshot of the system and monototic clock difference as a base reference
+	var smtDiffBase = sysMonTimeDiff()
+	for {
+		select {
+		case <-ticker.C:
+			// Check the IP change
+			currentIP, _ := util.GetHostIP()
+			if c.NodeIP != currentIP {
+				klog.Warningf("IP address has changed from %q to %q, restarting MicroShift", c.NodeIP, currentIP)
+				os.Exit(0)
```

Comment:
Does this trigger a clean shutdown, or do we need to do something to shutdown the subcomponents in a particular order? (If we need that, we can do it in a separate PR.)

Reply:
This code is a carry over from a previously approved PR. I do not know if this is the right way. I put it in our meeting agenda to discuss.